### PR TITLE
WAZO-783 provd registrars

### PR DESCRIPTION
**Pull request by bloom1**

### CODE REVIEW

File: integration_tests/suite/base/test_registrars.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+
+
+def test_create_registrar_null_parameters():
+    response = confd.registrars.post(
+        main_host='1.2.3.4',
+        proxy_main_host='1.2.3.4',
+        main_port=None,
+        name=None,
+        outbound_proxy_host=None,
+        outbound_proxy_port=None,
```
Comment (fblackburn1) [inline integration_tests/suite/base/test_registrars.py]: Why not setting all possible parameters to none? But I must admit that test is a little bit useless

File: integration_tests/suite/base/test_registrars.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+        'outbound_proxy_host': '1.2.3.8',
+        'outbound_proxy_port': 5064,
+    }
+
+    response = confd.registrars.post(**parameters)
+    response.assert_created('registrars')
+    assert_that(response.item, has_entries(parameters))
+
+    registrar = confd.registrars(response.item['id']).get()
+    assert_that(
```
Comment (fblackburn1) [inline integration_tests/suite/base/test_registrars.py]:
```suggestion
    assert_that(response.item, has_entries(parameters))
```

File: integration_tests/suite/base/test_registrars.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+        'proxy_backup_port': 5063,
+        'outbound_proxy_host': '1.2.3.8',
+        'outbound_proxy_port': 5064,
+    }
+
+    response = confd.registrars.post(**parameters)
+    response.assert_created('registrars')
+    assert_that(response.item, has_entries(parameters))
+
+    registrar = confd.registrars(response.item['id']).get()
```
Comment (fblackburn1) [inline integration_tests/suite/base/test_registrars.py]:
```suggestion
    response = confd.registrars(response.item['id']).get()
```
or
```suggestion
    registrar = confd.registrars(response.item['id']).get().item
```

File: integration_tests/suite/base/test_registrars.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+
+@fixtures.registrar(
+    main_host='1.2.3.4',
+    proxy_main_host='1.2.3.4',
+    main_port=5060,
+    name='TestRegistrar',
+    outbound_proxy_host='1.2.3.5',
+    outbound_proxy_port=5061,
+)
+def test_edit_registrar_null_parameters(registrar):
```
Comment (fblackburn1) [inline integration_tests/suite/base/test_registrars.py]: see comment test_post_registrar_null_parameters

File: wazo_confd/plugins/line/validator.py [hunk 5]
```diff
@@ -42,12 +50,12 @@ def build_validator(device_dao):
...
                 'registrar',
-                MemberOfSequence('registrar', device_dao.registrars, 'Registrar')
+                RegistrarExists(registrar_dao)
             ),
             GetResource('context', context_dao.get_by_name, 'Context'),
         ],
         edit=[
             ProvCodeChanged(line_dao),
-            MemberOfSequence('registrar', device_dao.registrars, 'Registrar'),
+            RegistrarExists(registrar_dao),
```
Comment (fblackburn1) [inline wazo_confd/plugins/line/validator.py]: Use `ResourceExists` validator

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,18 @@
...
+19.11
+-----
+
+* A new API for configuring the provisioning registrars has been added:
+
+  * GET `/1.1/registrars`
+  * POST `/1.1/registrars`
+  * GET `/1.1/registrars/<registrar_id>`
+  * PUT `/1.1/registrars/<registrar_id>`
+  * DELETE `/1.1/registrars/<registrar_id>`
```
Comment (fblackburn1) [inline CHANGELOG.md]: test

File: integration_tests/suite/helpers/helpers/registrar.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    ips = set(d['ip'] for d in response.items)
+    return _random_ip(ips)
+
+
+def generate_registrar(**params):
+    name = "".join(choice(string.ascii_letters) for _ in range(20))
+    ip = generate_ip()
+    registrar = {
+        'name': name,
+        'id': name,
```
Comment (fblackburn1) [inline integration_tests/suite/helpers/helpers/registrar.py]:
```suggestion
        'id': name,
```

File: wazo_confd/plugins/registrar/service.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        # Update all affected devices
+        for line in lines:
+            if line.device_id not in devices_updated:
+                self.device_updater.update_for_line(line)
+                devices_updated.add(line.device_id)
+
+        # Update default autoprov config
+        autoprov_config = self.provd_client.configs.get('autoprov')
+        autoprov_sip_line = autoprov_config['raw_config']['sip_lines']['1']
+        self._update_sip(resource.registrar, autoprov_sip_line)
```
[the remark this answers is not in the document]
Comment (bloom1) [reply wazo_confd/plugins/registrar/service.py]: Fixed

**Pull request merged**